src/xb-silo.h

```diff
@@ -97,4 +97,10 @@ gboolean	 xb_silo_get_enable_node_cache		(XbSilo		*self);
 void		 xb_silo_set_enable_node_cache		(XbSilo		*self,
 							 gboolean	 enable_node_cache);
 
+#include "xb-query.h"
+
+XbQuery		*xb_silo_lookup_query			(XbSilo		*self,
```

Comment:
do we need to export this?

Reply:
It’s not needed for any of the gnome-software or fwupd changes I’ve submitted so far, but I have a local gnome-software branch which uses it to cache the query instances which need values binding to them.

It basically exposes the query cache from the `XbSilo` publicly so that applications can get an `XbQuery` out of the silo to bind things to before querying using one of the new `query_with_bindings()` APIs. Those APIs all take an `XbQuery`, so the caller would otherwise either need to create their own cache, or create a new query each time with `xb_query_new()` (not performant).

Requiring that callers create and maintain their own cache is a viable alternative. However, when I tried it in gnome-software, it meant carrying the cache pointer around alongside the `XbSilo` everywhere, which was quite an invasive change.

I guess it depends on whether you think the `XbSilo` should, conceptually, have a cache for the queries it’s executed, or whether that should be separate.

There’s also a separate question about whether the `XbSilo` should cache *every* query it parses, or whether there should be a flag for callers to say “don’t cache the `XbQuery` which results from parsing this XPath string” because, say, the caller knows that query is never going to be used again.

Reply:
> it meant carrying the cache pointer around alongside the XbSilo everywhere

Ya, that sounds sucky.

> there should be a flag for callers to say “don’t cache the XbQuery

I very think we need this, else people shout at the memory use in fwupd again.

Reply:
> > there should be a flag for callers to say “don’t cache the XbQuery
> 
> I very think we need this, else people shout at the memory use in fwupd again.

I just looked at doing this, and I don’t think it will work: in all the APIs where the query is provided as an XPath string, there is no flags argument. In all the other cases, the caller is providing an `XbQuery`, so might as well explicitly call `xb_silo_lookup_cache()` if they want to use the cache.

How does it sound for me to disable caching by default in this MR, and amend the gnome-software/fwupd MRs to use `xb_silo_lookup_cache()` where queries need caching (and continue to use `xb_query_new()` otherwise).

Reply:
I’ve pushed changes to fix the other review comments, but I won’t address this one until you’ve replied on it, since it’s a more involved change.

Reply:
> in all the APIs where the query is provided as an XPath string, there is no flags argument

Is there any way to encode flags into the query string? It's not awesome, but what about something like `/components/component#cache=true`? Or we set the default for the _silo_?

> How does it sound for me to disable caching by default in this MR, and amend the gnome-software/fwupd MRs to use xb_silo_lookup_cache() where queries need caching (and continue to use xb_query_new() otherwise).

Yes, that works too. At least from a fwupd point of view it's sitting there idle 99.99% of the time and RSS usage is more important than saving a few ms on startup.

Reply:
OK, I’ve disabled caching by default.

I’ve also pushed a change which drops the `GError**` argument from `xb_silo_lookup_cache()`, on the basis that XPath queries should always be valid. This significantly simplifies the error handling in gnome-software (see the commit message for 2b4f58c23bc53c9940efc5eaa81561ebb7574c8b).

Ready for review again. My recent changes are all pushed as individual fixups, so you can easily see what’s changed since the last review.

Reply:
Aaargh, I’ve just had a horrible thought that `xb_query_set_flags()` and `xb_query_set_limit()` aren’t thread-safe either. If a cached query is simultaneously run from two different threads, either of those functions could be called with different values from each thread.

I think I might need to rework `XbValueBindings` to be `XbQueryContext` and to contain the flags and limit after all.

Watch this space.

Reply:
OK, updated. I’ve kept `XbValueBindings` (since they’re needed for the `XbMachine`), and added an `XbQueryContext` which wraps them in all the query APIs, and also contains the limit and flags. I’ve added an additional commit which deprecates `xb_query_{get,set}_{limit,flags}()`.

Ready for review again!

Reply:
I think this version is much better; good catch on the xb_query_set_flags() thread safety issue. I'm a bit confused why we need the QueryContext (which I like) and also the XbValueBindings object too. Don't they have the same lifetime and basic purpose?

Reply:
> Don't they have the same lifetime and basic purpose?

Not quite. The `XbQueryContext` is associated with a query, but the `XbValueBindings` is associated with a query and the `XbMachine` which runs for it. The `XbValueBindings` are copied/subsetted in a few places as different parts of the query are run. In my first attempt at this last night, `XbQueryContext` and `XbValueBindings` were the same object, and passing it (including the limit and flags) down to `XbMachine` was an awkward layering violation.

Reply:
> down to XbMachine was an awkward layering violation

Fair enough, thanks for the explanation. Can we put some of that text in the code for my future self please? Ta.

Reply:
Pushed another fixup commit to add a comment about that.